File: src/main/java/catmoe/fallencrystal/akanelimbo/rule/RuleHandler.java
package catmoe.fallencrystal.akanelimbo.rule;

import net.md_5.bungee.api.ProxyServer;
import net.md_5.bungee.api.config.ServerInfo;
import net.md_5.bungee.api.connection.ProxiedPlayer;
import net.md_5.bungee.api.event.ServerSwitchEvent;
import net.md_5.bungee.api.plugin.Listener;
import net.md_5.bungee.event.EventHandler;

import java.util.Timer;
import java.util.TimerTask;

public class RuleHandler implements Listener {

    ServerInfo switchfrom = ProxyServer.getInstance().getServerInfo("LoginLimbo");
    ServerInfo logined = ProxyServer.getInstance().getServerInfo("MainLimbo");

    @EventHandler
    public void ServerSwitchHandler(ServerSwitchEvent e) {
        ProxiedPlayer p = e.getPlayer();
        ServerInfo target = p.getServer().getInfo();
        try {
            // 从其它服务器跳转
            if (e.getFrom().equals(switchfrom) && target.equals(logined)) {
                Trigger(p);
            }
            // 使用try方法捕获null 表明玩家直接连接到该服务器
        } catch (NullPointerException ex) {
            // 如果目标是已经登录的服务器
            if (target.equals(logined)) {
                Trigger(p); // 触发
            }
        }
    }

    public void Trigger(ProxiedPlayer p) {
        if (hasBypassPermission(p)) {
            Skip(p);
            return;
        }
        Timer timer = new Timer();
        timer.schedule(new TimerTask() {
            @Override
            public void run() {
                try {
                    RuleMenu menu = new RuleMenu();
                    menu.close = false;
                    menu.open(p);
                } catch (NullPointerException ignore) {}
            }
        }, 1500);
    }

    public boolean hasBypassPermission(ProxiedPlayer p) {
        return p.hasPermission("rule.bypass");
    }

    public void Skip(ProxiedPlayer p) {
        ServerInfo lobby = ProxyServer.getInstance().getServerInfo("Lobby-1");
        p.connect(lobby);
    }
}

File: src/main/java/catmoe/fallencrystal/akanelimbo/rule/RuleHandler.kt
package catmoe.fallencrystal.akanelimbo.rule

import catmoe.fallencrystal.akanelimbo.StringManager
import catmoe.fallencrystal.moefilter.api.event.EventListener
import catmoe.fallencrystal.moefilter.api.event.FilterEvent
import catmoe.fallencrystal.moefilter.api.event.events.bungee.AsyncServerSwitchEvent
import net.md_5.bungee.api.connection.ProxiedPlayer
import java.util.*
import kotlin.concurrent.schedule

class RuleHandler : EventListener {
    private var loginLimbo = StringManager.getLoginLimbo()
    private var mainLimbo = StringManager.getMainLimbo()

    @FilterEvent
    fun serverSwitchHandler(e: AsyncServerSwitchEvent) {
        val p = e.player
        val target = p.server.info
        try {
            // 直接连接
            if (e.from == null && target == mainLimbo) { trigger(p) }
            // 从其它服务器跳转
            if (e.from == loginLimbo && target == mainLimbo) { trigger(p) }
        } catch (_: NullPointerException) { skip(p) }
    }

    private fun trigger(p: ProxiedPlayer) {
        // Enabled or Disabled
        if (!StringManager.getEnableRule()) {skip(p); return }
        // Cached read user from ReadCache.kt
        if (ReadCache.cacheGet(p.uniqueId) == true) {skip(p); return }
        val menu = RuleMenu()
        menu.closed = false
        menu.countdown(p)
        Timer().schedule(1500L) {
            try { menu.open(p) } catch (_: NullPointerException) { } catch (e: Exception) { skip(p) }
        }
    }

    private fun skip(p: ProxiedPlayer) { p.connect(StringManager.getLobby()) }
}
